bpf/web.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This BPF program is intended as an example for a simple web server
+ * that runs the services HTTP, HTTPS, SSH, and FTP.
+ *
+ * This BPF program builds upon the BPF program grantedv2, so
+ * there are primary and secondary limits. The secondary limit is only used
+ * for fragmented packets.
+ */
+
+#include <netinet/ip_icmp.h>
+#include <netinet/tcp.h>
+
+#include "grantedv2.h"
+
+#ifdef ntohs
+#undef ntohs
+#endif
+#define ntohs(x) __builtin_bswap16(x)
+
+SEC("init") uint64_t
+web_init(struct gk_bpf_init_ctx *ctx)
+{
+	return grantedv2_init_inline(ctx);
+}
+
+SEC("pkt") uint64_t
+web_pkt(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct grantedv2_state *state =
+		(struct grantedv2_state *)pkt_ctx_to_cookie(ctx);
+	struct rte_mbuf *pkt = pkt_ctx_to_pkt(ctx);
+	uint32_t pkt_len = pkt->pkt_len;
+	struct tcphdr *tcp_hdr;
+	uint64_t ret;
```

Comment:
Style: we can combine with the following line of code, i.e., ``uint64_t ret = grantedv2_pkt_begin(ctx, state, pkt_len);``